diff --git a/workbench_moderation.views.inc b/workbench_moderation.views.inc
index 9512698..9377ebd 100644
--- a/workbench_moderation.views.inc
+++ b/workbench_moderation.views.inc
@@ -14,9 +14,8 @@ function workbench_moderation_views_data() {
   $data['workbench_moderation_node_history']['table']['group'] = t('Workbench Moderation');
   $data['workbench_moderation_node_history']['table']['join'] = array(
     'node' => array(
-      'left_table' => 'node_revision',
-      'left_field' => 'vid',
-      'field' => 'vid',
+      'left_field' => 'nid',
+      'field' => 'nid',
       'type' => 'INNER',
     ),
     'node_revision' => array(