diff --git a/entityreference.module b/entityreference.module
index bdcb562..1d3a5f0 100644
--- a/entityreference.module
+++ b/entityreference.module
@@ -1026,7 +1026,7 @@ function entityreference_autocomplete_callback_get_matches($type, $field, $insta
     $entity = entity_load_single($entity_type, $entity_id);
     $has_view_access = (entity_access('view', $entity_type, $entity) !== FALSE);
     $has_update_access = (entity_access('update', $entity_type, $entity) !== FALSE);
-    if (!$entity || !($has_view_access || $has_update_access)) {
+    if (!$entity) {
       return MENU_ACCESS_DENIED;
     }
   }
