From 824e88c76c9b7bfe87cc2a6ea9625a78c52ae231 Mon Sep 17 00:00:00 2001
From: lucashedding <lucashedding@1463982.no-reply.drupal.org>
Date: Wed, 26 Nov 2014 15:45:39 -0600
Subject: [PATCH] Issue #2382855 by heddn: Required field collection, all
 optional child fields

---
 field_collection.module | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/field_collection.module b/field_collection.module
index 3a8f9e9..dde2e7a 100644
--- a/field_collection.module
+++ b/field_collection.module
@@ -1763,6 +1763,16 @@ function field_collection_field_widget_embed_validate($element, &$form_state, $c
       }
     }
   }
+  // If the field collection is required but has all optional fields,
+  // require the field collection element itself.
+  elseif($element['#required'] && field_collection_item_is_empty($field_collection_item)) {
+    if (!empty($field_state['instance']['label'])) {
+      form_error($element, t('!name field is required.', array('!name' => $field_state['instance']['label'])));
+    }
+    else {
+      form_error($element);
+    }
+  }
 
   // Only if the form is being submitted, finish the collection entity and
   // prepare it for saving.
-- 
2.2.0

