diff --git a/i18n_taxonomy/i18n_taxonomy.inc b/i18n_taxonomy/i18n_taxonomy.inc
index 104f0c0..ba5c454 100644
--- a/i18n_taxonomy/i18n_taxonomy.inc
+++ b/i18n_taxonomy/i18n_taxonomy.inc
@@ -47,6 +47,6 @@ class i18n_taxonomy_term extends i18n_string_object_wrapper {
    * Access to object translation. This should check object properties and permissions
    */
   protected function translate_access() {
-    return taxonomy_term_edit_access($this->object) && $this->get_langcode() && user_access('translate interface');
+    return taxonomy_term_edit_access($this->object) && $this->get_langcode() && user_access('translate taxonomies');
   }
 }
diff --git a/i18n_taxonomy/i18n_taxonomy.module b/i18n_taxonomy/i18n_taxonomy.module
index d5c1793..df3719c 100644
--- a/i18n_taxonomy/i18n_taxonomy.module
+++ b/i18n_taxonomy/i18n_taxonomy.module
@@ -117,6 +117,15 @@ function i18n_taxonomy_menu() {
   return $items;
 }
 
+function i18n_taxonomy_permission() {
+  return array(
+    'translate taxonomies' =>  array(
+      'title' => t('Translate taxonomies and their terms'),
+      'restrict access' => TRUE,
+    ),
+  );
+}
+
 /**
  * Implements hook_admin_paths().
  */
@@ -164,7 +173,7 @@ function i18n_taxonomy_vocabulary_translation_tab_sets_access($vocabulary) {
  * @todo This should work also for localizable terms when we've got that part implemented
  */
 function i18n_taxonomy_term_translation_tab_access($term) {
-  return taxonomy_term_edit_access($term) && i18n_taxonomy_vocabulary_mode($term->vid) & I18N_MODE_MULTIPLE && user_access('translate interface');
+  return taxonomy_term_edit_access($term) && i18n_taxonomy_vocabulary_mode($term->vid) & I18N_MODE_MULTIPLE && user_access('translate taxonomies');
 }
 
 /**
@@ -702,7 +711,7 @@ function i18n_taxonomy_form_taxonomy_form_vocabulary_alter(&$form, &$form_state)
     $langcode = i18n_object_langcode($vocabulary, LANGUAGE_NONE);
     // Define the replacement names to add some logic to the translation mode options.
     $form += i18n_translation_mode_element('taxonomy_vocabulary', $i18n_mode, $langcode);
-    if (user_access('translate interface')) {
+    if (user_access('translate taxonomies')) {
       $form['actions']['translate'] = array(
         '#type' => 'submit',
         '#name'   => 'save_translate',
@@ -833,7 +842,7 @@ function i18n_taxonomy_form_taxonomy_form_term_alter(&$form, &$form_state) {
       break;
   }
 
-  if (user_access('translate interface') && i18n_taxonomy_vocabulary_mode($vocabulary->vid) & I18N_MODE_MULTIPLE) {
+  if (user_access('translate taxonomies') && i18n_taxonomy_vocabulary_mode($vocabulary->vid) & I18N_MODE_MULTIPLE) {
     $form['actions']['translate'] = array(
       '#type' => 'submit',
       '#name'   => 'save_translate',
