--- field_collection_node_clone/field_collection_node_clone.module	(revision 141)
+++ field_collection_node_clone/field_collection_node_clone.module	(revision 145)
@@ -46,11 +46,11 @@
   $entity_wrapper = entity_metadata_wrapper($entity_type, $entity);
   $old_items = $entity_wrapper->{$fc_field}->value();
 
+  // This is a fix to fill the $old_item variables when it's empty.
+  if (empty($old_items)) {
+    foreach ($entity->{$fc_field}[$language] as $item_key => $item_value) {
+      $fc_item = field_collection_item_load($item_value['value']);
+      $old_items[$item_key] = $fc_item;
+    }
+  }
